[Python][CI] test_table_uses_memory_pool flaky on macOS 14 job

### Describe the bug, including details regarding any error messages, version, and platform.

`test_table_uses_memory_pool` fails occasionally with `.../test_pandas.py:3964: AssertionError`
Observed on job `ARM64 macOS 14`, where `pytest -r s -n auto ...` uses xdist to run the tests in parallel (8207, so ~2700 per worker).
[Latest example May 25](https://github.com/apache/arrow/actions/runs/26399836679/job/77709190993?pr=50033#step:10:210), same commit [re-run passed](https://github.com/apache/arrow/actions/runs/26399836679/job/77727866997?pr=50033#step:10:208).
[Second example May 20](https://github.com/apache/arrow/actions/runs/26191758355/job/77061638937#step:10:230)
[Third example May 20 is on main](https://github.com/apache/arrow/actions/runs/26135872194/job/76870837319#step:10:230)

```bash
    def test_table_uses_memory_pool():
        N = 10000
        arr = pa.array(np.arange(N, dtype=np.int64))
        t = pa.table([arr, arr, arr], ['f0', 'f1', 'f2'])
    
        prior_allocation = pa.total_allocated_bytes()
...
        x = None  # noqa
        gc.collect()
>       assert pa.total_allocated_bytes() == prior_allocation
E       assert 1472 == 1664
...
```

Resembles #44728: `prior_allocation` captures before `gc.collect()`, so lingering allocations from earlier tests in the same xdist worker could be counted in the baseline but later freed by `gc.collect()` making final smaller than baseline.

### Component(s)

Python, Continuous Integration

--- a/python/pyarrow/tests/test_pandas.py
+++ b/python/pyarrow/tests/test_pandas.py
@@ -3966,6 +3966,7 @@ def test_table_uses_memory_pool():
     arr = pa.array(np.arange(N, dtype=np.int64))
     t = pa.table([arr, arr, arr], ['f0', 'f1', 'f2'])
 
+    gc.collect()
     prior_allocation = pa.total_allocated_bytes()
     x = t.to_pandas()
 

--- a/python/pyarrow/tests/test_pandas.py
+++ b/python/pyarrow/tests/test_pandas.py
@@ -3901,6 +3901,7 @@ def test_singleton_blocks_zero_copy():
 
 
 def _check_to_pandas_memory_unchanged(obj, **kwargs):
+    gc.collect()
     prior_allocation = pa.total_allocated_bytes()
     x = obj.to_pandas(**kwargs)  # noqa
 

